### ISSUE TITLE
C/C++ include directives and explicitness

### ISSUE THREAD

**Issue by PeterBowman** [roboticslab-uc3m/best-practices#19]
From https://github.com/roboticslab-uc3m/best-practices/issues/12#issuecomment-337389502:

> I just want to leave this here to illustrate a case where minimizing `#include`s is hazardous (which is vaguely related to this issue, but still...). Compare these two CI weekly cron builds of the [tools repository](https://github.com/roboticslab-uc3m/tools/) on consecutive commits, no changes in between:
>
> * https://travis-ci.org/roboticslab-uc3m/tools/builds/286195407
> * https://travis-ci.org/roboticslab-uc3m/tools/builds/289191356
>
> One cannot depend on headers that may or may not include another header, which in turn happens to be a must for your code to compile, and that's why I tend to prefer explicitness. This was the fix: roboticslab-uc3m/tools@da3ee49.

I think this is worth a mention in our guides. Two more situations I recently ran into:

* [missing <sys/ioctl.h>](https://github.com/roboticslab-uc3m/yarp-devices/commit/b579638700a312b0c8f88e066afa575a94b06ebc#diff-b191eef97b53d8bf49a9778b58e8ffbd): another transitive inclusion during a major refactorization effort which had nothing to do with the Jr3 device
* `<yarp/dev/CanBusInterface.h>` is (surprise!) not included in `<yarp/dev/all.h>`

The latter means that we may want to avoid omnibus headers of the sort of `xxx/all.h`. Is this acceptable?

Also reflect conclusions of #12.

**Comment by jgvictores** [comment]
I totally agree on advising against omnibus headers of the sort of `xxx/all.h`. :+1:

**Comment by PeterBowman** [comment]
More refs, now focusing on self-sufficient headers:

* [Self-sufficient header files in C/C++ (SO)](https://stackoverflow.com/questions/1892043/self-sufficient-header-files-in-c-c)
* [Self-Sufficient Headers (Herb Sutter)](http://www.drdobbs.com/self-sufficient-headers/184401705)
* https://github.com/robotology/yarp/issues/1641

**Comment by PeterBowman** [comment]
> `<yarp/dev/CanBusInterface.h>` is (surprise!) not included in `<yarp/dev/all.h>`

I presume this could be treated as an upstream bug. Nevertheless, good practice (and common sense?) dictates avoidance of omnibus header abuse. Another example is the `IControlLimits2.h` header (deprecated in YARP 3.0), which looks odd just beside the `<yarp/dev/all.h>` inclusion: https://github.com/roboticslab-uc3m/yarp-devices/commit/96003c6ed58ca4c05bbff2cfc3d79f9a229ed9af.

**Comment by PeterBowman** [comment]
Looking at this the other way around, i.e. redundant/unnecessary included headers, the [`include-what-you-use` tool](https://include-what-you-use.org/) seems an interesting approach: [SO question](https://stackoverflow.com/q/30951492).

**Comment by PeterBowman** [comment]
Someone may be tempted to give for granted the transitiveness of included headers under the assumption that usage requirements, once being reflected in the public API, are automatically propagated. Example:

```c++
// file "FancyInterface.hpp"
#include <SomeFancyDependency.hpp>

class FancyInterface
{
public:
    void doSomething(SomeFancyDependency & obj),
};
```

Let's say I'm an user of this interface and I want to invoke an implementation of *FancyInterface::doSomething*. The signature tells me it introduces the *SomeFancyDependency*. I am pretty safe to include *FancyInterface.hpp* and assume everything will work.

Now, see what happens in the following example:

```c++
// file "FancyInterface.hpp"
class SomeFancyDependency; // forward declaration

class FancyInterface
{
public:
    void doSomething(SomeFancyDependency * p_obj),
};
```

Since `SomeFancyDependency` is referred to by pointer, there is no need to provide further hints to the compiler apart from a forward declaration. This is a so-called [incomplete type](https://en.cppreference.com/w/cpp/language/type#Incomplete_type) and can be found e.g. in [yarp::os::PortReader](https://github.com/robotology/yarp/blob/24a58596bbcab211cf763ad179716636b6cd61c7/src/libYARP_os/src/yarp/os/PortReader.h). Now, if you implement *FancyInterface::doSomething* and fail to include *SomeFancyDependency.hpp* **by yourself**, the compiler will complain loudly.

**Comment by PeterBowman** [comment]
Also affects [`yarp::os::TypedReaderCallback`](http://www.yarp.it/git-master/classyarp_1_1os_1_1TypedReaderCallback.html). Implementors need to include [`yarp::os::TypedReader`](http://www.yarp.it/git-master/classyarp_1_1os_1_1TypedReader.html).
